docs/developer/architecture/user_customization_and_persistence.rst

```diff
@@ -0,0 +1,225 @@
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly.
+* ConfigMap data is store in the :term:`etcd` database which is generally being
+  backed up. This ensures that user settings cannot be lost easily.
+
+.. note::
+
+   Persisting newly added Grafana dashboards or new Grafana datasources
+   especially for modifications added via the Grafana UI cannot be stored in
+   ConfigMaps.
+   To handle this particular user case, there is a need to provision persistent
+   storage volumes to handle the persistence of these settings across Pod
+   restarts.
+
+Rejected design choices
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Consul KV
+~~~~~~~~~
+
+This approach offers a full fledge KV store with a /kv endpoint which allows
+CRUD operations on all KV data stored in it.
+Consul KV also allows access to past versions of objects and has an optimistic
+concurrency when manipulating multiple objects.
+
+Note that, Consul KV store was rejected because managing operations such as
+performing full backups, system restores for a full fledged KV system
+requires time and much more efforts than the ConfigMap KV store which is
+simplistic and matches the requirements stated.
+
+
+Implementation Details
+----------------------
+
+Storage format
+~~~~~~~~~~~~~~
+
+A sample ConfigMap store can be defined with the following fields.
+
+An example of such a ConfigMap store:
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      namespace: <namespace>
+      name: <config-name>
+    data:
+      config.yaml: |-
+        apiVersion: <object-version>
+        kind: <kind>
+        spec:
+          <key>: <values>
+
+**Use case 1:**
+
+Configure and store the number of replicas for service specific Deployments
+found in the `metalk8s-monitoring` namespace using the ConfigMap store format.
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      namespace: metalk8s-monitoring
+      name: metalk8s-grafana-userconfig
+    data:
+      config.yaml: |-
+        apiVersion: metalk8s.scality.com/v1alpha1
+        kind: GrafanaUserConfig
+        spec:
+          replicas: 2
+
+How it works
+~~~~~~~~~~~~
+
+Service pods and deployments will be configured to consume configuration data
+directly from their respective minion external pillars.
+
+During Bootstrap, these external pillar values will be pre-filled with default
+values and the service consumers will be configured to use these values.
+
+**Using Saltstates**
+
+Once a ConfigMap KV store is updated by the user(say a user changes the
+number of replicas for Prometheus deployments to a new value), then perform the
+following actions;
+
+  - Apply a salt state that reads the ConfigMap object, validates the schema
+    based on MetalK8s defined standards and checks the new values passed.
+  - If the ConfigMap object is valid, the new values passed by the user are
+    then re-rendered to the pillars.
```

Comment:
Not sure we really want this value to be in the pillar, IMO just reading the CM directly in the state is better so that user can NOT change, for example, the number of replicas just on CLI using `pillar="blabla"` but he always need to edit the CM

Reply:
Resolved